roadmap/implementers-guide/src/node/availability/availability-recovery.md

```diff
@@ -20,3 +22,148 @@ Output:
+    received_chunks: Map<ValidatorIndex, ErasureChunk>,
+    requesting_chunks: FuturesUnordered<Receiver<ChunkResponse>>,
+}
+```
+
+### Signal Handling
+
+On `ActiveLeavesUpdate`, if `activated` is non-emtpy, set `state.live_block_hash` to the first block in `Activated. Also, remove any items that have been in 
+
+Ignore `BlockFinalized` signals.
+
+On `Conclude`, shut down the subsystem.
+
+#### `AvailabilityRecoveryMessage::RecoverAvailableData(receipt, session, response)`
+
+1. Check the `availability_lru` for the candidate and return the data if so.
+1. Check if there is already an interaction handle for the request. If so, add the response handle to it.
+1. Otherwise, load the session info for the given session under the state of `live_block_hash`, and initiate an interaction with *launch_interaction*. Add an interaction handle to the state and add the response channel to it.
+1. If the session info is not available, return `RecoveryError::Unavailable` on the response channel.
+
+### From-interaction logic
+
+#### `FromInteraction::Concluded`
+
+1. Load the entry from the `interactions` map. It should always exist, if not for logic errors. Send the result to each member of `awaiting`.
+1. Add the entry to the availability_lru.
+
+#### `FromInteraction::MakeRequest(discovery_pub, candidate_hash, validator_index, response)`
+
+1. Add an `AwaitedRequest` to the `discovering_validators` map under `discovery_pub`.
+1. Issue a `NetworkBridgeMessage::ConnectToValidators`.
+1. Add the stream of connected validator events to `state.connecting_validators`.
+
+#### `FromInteraction::ReportPeer(peer, rep)`
+
+1. Issue a `NetworkBridgeMessage::ReportPeer(peer, rep).
+
+### Responding to network events.
+
+#### On `connecting_validators` event:
+
+1. If the validator exists under `discovering_validators`, remove the entry.
+1. For each `AwaitedChunk` in the entry, issue a `AvailabilityRecoveryV1Message::RequestChunk(next_request_id, candidate_hash, validator_index)` and make an entry in the `live_chunk_requests` map.
+
+#### On receiving `AvailabilityRecoveryV1::RequestChunk(r_id, candidate_hash, validator_index)`
+
+1. Issue a `AvailabilityStore::QueryChunk(candidate-hash, validator_index, response)` message.
+1. Whatever the result, issue a `AvailabilityRecoveryV1Message::Chunk(r_id, response)` message.
+
+#### On receiving `AvailabilityRecoveryV1::Chunk(r_id, chunk)`
+
+1. If there exists an entry under `r_id`, remove it. If there doesn't exist one, report the peer and return. If the peer in the entry doesn't match the sending peer, reinstate the entry, report the peer, and return.
+1. Send the chunk response on the `awaited_chunk` for the interaction to handle.
+
+### Interaction logic
+
+#### `launch_interaction(session_index, session_info, candidate_receipt, candidate_hash)`
+
+1. Compute the threshold from the session info. It should be `f + 1`, where `n = 3f + k`, where `k in {1, 2, 3}`, and `n` is the number of validators.
```

Comment:
DW: I always believed that this threshold in the AV protocol simply put says this: "as long as 2/3+1 validators are honest and will send us the piece the recovery would work. So if we request a piece from every validator at least 2/3+1 would send us a chunk and then this will be enough to recover data". Also `inclusion.rs` defines  the threshold as

```rust
const fn availability_threshold(n_validators: usize) -> usize {
	let mut threshold = (n_validators * 2) / 3;
	threshold += (n_validators * 2) % 3;
	threshold
}
```

So two questions here:
 1. Which is the correct formula for the threshold?
 1. We should probably be querying all validators and waiting for at least `threshold` correct pieces shouldn't we?

Reply:
1. We only require `f + 1` (1/3 + 1) pieces to recover from the erasure coding because that is a subset which is guaranteed to contain at least one honest node's piece. However, we consider something available (by the bitfield system) only when `2f + 1` validators claim that they have their piece, because this means that there is at least a set of `f + 1` honest nodes who all have their piece, and thus the data will still be available even if all `f` adversarial nodes refuse to respond to requests.

2. That's effectively what the `Interaction` does. It queries random batches of validators for the session, a few at a time, and asks them each for their piece. Until we have f+1 pieces, and then we attempt to recover.